flamingpy/noise/cv.py

```diff
@@ -0,0 +1,649 @@
+
+    def _permute_indices_and_label(self):
+        """Obtain permuted indices and set type of reduced node.
+
+        For each macronode, permute indices so that the first encountered GKP state
+        comes first, designating it as the 'star' ('central') mode. The first
+        index in the resulting list and every four indices thereafter correspond
+        to star modes. The rest are 'planets' ('satellite' modes).
+
+        For each star, associate a 'body_index' of 1, and 2, 3, and 4 for the
+        subsequent planets. Additionally, if a macronode contains at least one GKP
+        state, label the reduced state as 'GKP' (otherwise 'p').
+
+        This method sets the attribute self.permuted_inds to the permuted
+        indices and modifies self.egraph and self.reduced_graph.
+        """
+        N = self._N
+        macro_graph = self.egraph
+        to_points = macro_graph.to_points
+        # A list of permuted indices where each block of four
+        # corresponds to [star, planet, planet, planet].
+        permuted_inds = np.empty(N, dtype=np.int32)
+        for i in range(0, N - 3, 4):
+            # Indices of GKP micronodes in macronode i.
+            gkps = []
+            for j in range(4):
+                micronode = to_points[i + j]
+                if macro_graph.nodes[micronode]["state"] == "GKP":
+                    gkps.append(j)
+            centre_point = tuple(round(i) for i in micronode)
+            if gkps:
+                star_ind, reduced_state = i + gkps[0], "GKP"
+            else:
+                star_ind, reduced_state = i, "p"
+            # Set type of node in the reduced lattice as a p-squeezed
+            # state if all micronodes are p, else GKP.
+            self.reduced_graph.nodes[centre_point]["state"] = reduced_state
+            # Old and permuted indices of all micronodes in macronode i.
+            old_inds = [i, i + 1, i + 2, i + 3]
+            old_inds.pop(star_ind - i)
+            new_inds = [star_ind] + old_inds
+            # Associate a 'body index' (1 to 4) to each micronode,
+            # with 1 being the star index and the rest being planets.
+            k = 1
+            for ind in new_inds:
+                macro_graph.nodes[to_points[ind]]["body_index"] = k
+                k += 1
+            permuted_inds[[i, i + 1, i + 2, i + 3]] = new_inds
+        self.permuted_inds = permuted_inds
+
+    def _entangle_states(self):
+        """Entangle the states in the macro_graph.
+
+        Apply CZ gates to (i.e. a symplectic CZ matrix to the quadratures of)
+        self.egraph, based on where the edges are in the graph. Then, apply the
+        four-splitter to each macronode.
+
+        This method sets the attributes self.permuted_quads to the permuted
+        quadratures and self.quad_permutation to the corresponding permutation
+        vector.
+        """
+        macro_graph = self.egraph
+        quads = SCZ_apply(macro_graph.adj_mat, self._init_quads)
```

Comment:
This was already computed.

Reply:
Where?

Reply:
In the call to _apply_noise.

Reply:
Anyway, this point is now moot with some of the latest changes.